dali/operators/decoder/audio/audio_decoder.h

```diff
@@ -47,10 +47,34 @@ class AudioDecoderBase {
   }
 
   /**
-   * @brief Decode audio data and store it in the supplied buffer
-   * @return Number of (multi-channel) samples actually read
+   * @brief Determines whether the particular decoder can seek frames
    */
-  virtual ptrdiff_t Decode(span<char> raw_output) = 0;
+  virtual bool CanSeekFrames() const = 0;
+
+  /**
+   * @brief Seeks full frames, or multichannel samples, much like lseek in unistd.h
+   * @param nframes Number of full frames (1 frame is equivalent to nchannel samples)
+   * @param whence Like in lseek, SEEK_SET. SEEK_CUR, SEEK_END
+   */
+  virtual int64_t SeekFrames(int64_t nframes, int whence = SEEK_CUR) = 0;
+
+  /**
+   * @brief Decode audio samples.
+   * @remarks output length should include the number of channels
+   *          (audio_length * num_channels)
+   * @return Number of samples read
+   */
+  virtual ptrdiff_t Decode(span<float> output) = 0;
+  virtual ptrdiff_t Decode(span<int16_t> output) = 0;
+  virtual ptrdiff_t Decode(span<int32_t> output) = 0;
```

Comment:
No me gusta :(
What's the reason for removing `TypedDecoder` and having these 3 virtuals here? It feels like the opposite of how the interface should look like...

Reply:
TypedDecoder introduced a compile time limitation that was artificial. If instantiate a float decoder because I think that I will need to do resampling, but later I get a recording with the same sampling rate as the requested one, I couldn't decode directly to T because of the static type of the instance.

Reply:
Can't agree more with Joaquin. If an interface gets in the way of functionality, it's not a good one.

Reply:
One more note, defining the data type at instantiation introduces this problem, because it is only after I have read the file meta-data that I can decide the data type of the output.

Reply:
This interface, however, exposes the implementation detail: that there are different ways to decode different types. I'd much more prefer to hide it:
```c++
public:
template<typename T>
ptrdiff_t Decode (span<T> output) {
  if (is_type<T, float>) return DecodeImpl(output);
  [...]
  else  DALI_FAIL("Type not supported");

}

private:
virtual ptrdiff_t DecodeImpl(span<float> output) = 0;
virtual ptrdiff_t DecodeImpl(span<int16_t> output) = 0;
virtual ptrdiff_t DecodeImpl(span<int32_t> output) = 0;
```

Reply:
How about hiding the implementation detail, that there are different ways to decode different types? Something like we do for `Open` and `Close`:
```c++
public:
template<typename T>
ptrdiff_t Decode (span<T> output) {
  return DecodeImpl(output);
}

private:
virtual ptrdiff_t DecodeImpl(span<float> output) = 0;
virtual ptrdiff_t DecodeImpl(span<int16_t> output) = 0;
virtual ptrdiff_t DecodeImpl(span<int32_t> output) = 0;
```
I'm thinking in the way that if we would need, e.g. perform some particular check before actually decoding. This way we would do it in a single point, not in every function with implementation:
```c++
ptrdiff_t Decode(span<T> output) {
  assert(output.length > 0);
  return DecodeImpl(output);
}
```
What bothers me in this interface how it looks now is that it exposes that implementation detail...